src/cascadia/TerminalSettingsEditor/Appearances.cpp

```diff
@@ -42,6 +43,157 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         return _hasPowerlineCharacters.value_or(false);
     }
 
+    Windows::Foundation::Collections::IMap<winrt::hstring, winrt::hstring> Font::FontAxesTagsAndNames()
+    {
+        if (!_fontAxesTagsAndNames)
+        {
+            _fontAxesTagsAndNames = winrt::single_threaded_map<winrt::hstring, winrt::hstring>();
+
+            wil::com_ptr<IDWriteFont> font;
+            THROW_IF_FAILED(_family->GetFont(0, font.put()));
+            wil::com_ptr<IDWriteFontFace> fontFace;
+            THROW_IF_FAILED(font->CreateFontFace(fontFace.put()));
+            wil::com_ptr<IDWriteFontFace5> fontFace5;
+            if (fontFace5 = fontFace.try_query<IDWriteFontFace5>())
+            {
+                wil::com_ptr<IDWriteFontResource> fontResource;
+                THROW_IF_FAILED(fontFace5->GetFontResource(fontResource.put()));
+
+                std::vector<DWRITE_FONT_AXIS_VALUE> axesVector;
+                const auto axesCount = fontFace5->GetFontAxisValueCount();
+                if (axesCount > 0)
+                {
+                    uint32_t localeIndex;
+                    BOOL localeExists;
+                    wchar_t localeName[LOCALE_NAME_MAX_LENGTH];
+                    const auto localeToTry = GetUserDefaultLocaleName(localeName, LOCALE_NAME_MAX_LENGTH) ? localeName : L"en-US";
+
+                    axesVector.resize(axesCount);
+                    fontFace5->GetFontAxisValues(axesVector.data(), axesCount);
+                    for (uint32_t i = 0; i < axesCount; ++i)
+                    {
+                        const auto tagString = _axisTagToString(axesVector[i].axisTag);
+
+                        wil::com_ptr<IDWriteLocalizedStrings> names;
+                        THROW_IF_FAILED(fontResource->GetAxisNames(i, names.put()));
+
+                        if (!SUCCEEDED(names->FindLocaleName(localeToTry, &localeIndex, &localeExists)) || !localeExists)
+                        {
+                            // default to the first locale in the list
+                            localeIndex = 0;
+                        }
+
+                        uint32_t length;
+                        if (SUCCEEDED(names->GetStringLength(localeIndex, &length)))
+                        {
+                            // it is reasonable to assume that the name length is not going to exceed 512 chars
+                            wchar_t name[512];
+                            if (SUCCEEDED(names->GetString(localeIndex, name, length + 1)))
```

Comment:
This is not safe: You don't know whether `length` is actually <512. Just drop the `GetStringLength` call and pass `ARRAYSIZE(name)` as the length parameter. Alternatively, use this:

```cpp
static winrt::hstring getLocalizedStringByIndex(IDWriteLocalizedStrings* strings, UINT32 index)
{
    UINT32 length = 0;
    THROW_IF_FAILED(strings->GetStringLength(index, &length));

    winrt::impl::hstring_builder builder{ length };
    THROW_IF_FAILED(strings->GetString(index, builder.data(), length + 1));

    return builder.to_hstring();
}
```

Reply:
Heh the `512` upper limit was [your suggestion](https://github.com/microsoft/terminal/pull/16104#discussion_r1467079130)! Updated the PR with your new suggestion though